### PR TITLE
Show Featured Items

### PR DESCRIPTION
**Pull request by MuhweziDeo**
- Show Featured Items in Home page

### CODE REVIEW

File: geonode_mapstore_client/client/js/routes/Home.jsx [hunk 1]
```diff
@@ -422,6 +435,19 @@ function Home({
             <div className="gn-main-home">
 
                 <div className="gn-container">
+                    <div className="gn-row">
+                        <div className="gn-grid-container">
+                            <ConnectedFeatureList
+                                query={query}
+                                location={location}
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/routes/Home.jsx:442]: We don't need to pass the location to check if the component should be rendered or not but we could add a props boolean to the Home route to disable/enable the FeaturedList (eg: disableFeatured). In this way we can add disableFeatured equal to true in the Search and Detail routes components
```js
{!disableFeatured && <ConnectedFeatureList
```

File: geonode_mapstore_client/client/js/components/home/FeaturedList.jsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        ...(!isNextPageAvailable || loading ? {color: 'grey', cursor: 'not-allowed'} : {cursor: 'pointer'})
+    };
+
+    const previousIconStyles = {
+        fontSize: '2rem',
+        marginRight: '1rem',
+        ...(!isPreviousPageAvailable || loading ? {color: 'grey', cursor: 'not-allowed'} : {cursor: 'pointer'})};
+    return (
+        <>
+            {hasResources && location.pathname === '/' && <div className="gn-card-grid">
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/components/home/FeaturedList.jsx:113]: see previous comment about location

File: geonode_mapstore_client/client/js/epics/gnsearch.js [hunk 3]
```diff
@@ -225,9 +242,27 @@ export const gnsSelectResourceEpic = (action$, store) =>
                 );
         });
 
+export const getFeaturedResourcesEpic = (action$, {getState = () => {}}) =>
+    action$.ofType(LOCATION_CHANGE, UPDATE_FEATURED_RESOURCES)
+        .filter(({ payload }) => {
+            return payload ? payload.location?.pathname === '/' : getState().router?.location?.pathname === '/';
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/epics/gnsearch.js:248]: see previous comment about location

File: geonode_mapstore_client/client/js/routes/Home.jsx [hunk 4]
```diff
@@ -422,6 +435,19 @@ function Home({
             <div className="gn-main-home">
 
                 <div className="gn-container">
+                    <div className="gn-row">
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/routes/Home.jsx:438]: We should apply a body-bg color to this component so could add a new class called gn-home-section to this div

```html
<div className="gn-row gn-home-section">
```

This is the expected mixin to apply in the geonode.scss file
```scss
@include background-color-var(map.get($theme-vars, 'body-bg'), --gn-body-bg);
```

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

File: geonode_mapstore_client/client/js/epics/gnsearch.js (hunk 3 above)
Comment (MuhweziDeo) [reply geonode_mapstore_client/client/js/epics/gnsearch.js:248]: This is used to trigger fetching the resources on location change. Maybe I could dispatch the action to fetch the featured resources manually in the useEffect

File: geonode_mapstore_client/client/js/routes/Home.jsx [hunk 5]
```diff
@@ -341,6 +357,15 @@ function Home({
...
     }, []);
 
+    useEffect(() => {
+        if (!disableFeatured) {
+            const margin = 24;
+            const size = 320;
+            const count = Math.floor((width || window.outerWidth) / (size + margin));
+            fetchFeaturedResources(undefined, count);
+        }
+    }, [disableFeatured]);
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/routes/Home.jsx:367]: Just tried and this has not effect on the pagination, the count is always 5 in my case.
I think we need to move this directly inside the FeaturedList list component. So please remove this useEffect and see my next comment

File: geonode_mapstore_client/client/js/components/home/FeaturedList.jsx [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    isPreviousPageAvailable,
+    loadFeaturedResources,
+    detectedWidth,
+    containerWidth
+}) => {
+
+    const width = containerWidth || detectedWidth;
+    const margin = 24;
+    const size = 320;
+    const count = Math.floor(width / (size + margin));
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/components/home/FeaturedList.jsx:103]: In this case we are wrapping the component with withResizeDetector but not using width provided by this HOC. There is something that should be checked here. The width we need is this one:

image

and withResizeDetector should compute that value so please try to remove detectedWidth and containerWidth props and use width prop. We should also move the initial load of resources here and if the count changes we should trigger a new request 

```js
    width,
    onLoad
}) => {

    const margin = 24;
    const size = 320;
    const count = Math.floor(width / (size + margin));
    useEffect(() => {
        onLoad(undefined, count);
    }, [count]);
```

File: geonode_mapstore_client/client/js/routes/Home.jsx [hunk 7]
```diff
@@ -415,6 +440,20 @@ function Home({
             <div className="gn-main-home">
 
                 <div className="gn-container">
+                    <div className="gn-row gn-home-section">
+                        <div className="gn-grid-container">
+                            {!disableFeatured &&  <ConnectedFeatureList
+                                query={query}
+                                formatHref={handleFormatHref}
+                                buildHrefByTemplate={buildHrefByTemplate}
+                                detectedWidth={width}
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/routes/Home.jsx:449]: please remove detectedWidth and add onLoad={fetchFeaturedResources}

File: geonode_mapstore_client/client/js/components/home/FeaturedList.jsx [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+ */
+
+import React from 'react';
+import { Spinner, Button } from 'react-bootstrap-v1';
+import HTML from '@mapstore/framework/components/I18N/HTML';
+import ResourceCard from '@js/components/home/ResourceCard';
+import FaIcon from '@js/components/home/FaIcon';
+import { withResizeDetector } from 'react-resize-detector';
+
+const Cards = withResizeDetector(({
```
Comment (allyoucanmap) [inline geonode_mapstore_client/client/js/components/home/FeaturedList.jsx:16]: The width seems provided by the parent so the withResizeDetector is not needed here. Please check if removing this has some effect

**Review by allyoucanmap — CHANGES_REQUESTED** [review]

**Review by allyoucanmap — APPROVED** [review]

**Pull request merged**